### PR TITLE
[RFC] State Snapshotting

### PR DESCRIPTION
**Pull request by bkase**
RFC for state snapshotting

### CODE REVIEW

File: rfcs/0017-state-snapshotting.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
+## Summary
+[summary]: #summary
+
+This RFC provides information on taking state snapshots. These snapshots would
```
Comment (psteckler) [inline rfcs/0017-state-snapshotting.md]: Which piece of state is being snapshotted?

File: rfcs/0017-state-snapshotting.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+required above:
+
+The following is what is believed to be an exhaustive list of the pieces of
+state needed for full-offline snapshotting:
+
+1. Transition frontier + extensions
+
+For marshalling and unmarshalling of the data:
+
+We can rely on https://github.com/CodaProtocol/coda/pull/1779 (which should be
```
Comment (psteckler) [inline rfcs/0017-state-snapshotting.md]: should describe here what the subject of that PR is

File: rfcs/0017-state-snapshotting.md [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+group2:  B----------
+person1:            C-----
+person2:                  D--
+```
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+* These tasks would take quite a bit of effort to complete
+* The snapshotting system itself will be complex enough to have it's own bugs
```
Comment (enolan) [inline rfcs/0017-state-snapshotting.md]: wrong its

File: rfcs/0017-state-snapshotting.md [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+7. Transaction pool
+8. Snark pool
+
+When we want to start the node from the snapshot, we need to boot it into a
+special state that will execute the proposer to generate the next state and
+transition and then stall afterwards.
+
+For marshalling and unmarshalling and copying the data:
+
+* `bin_io` can be trivally used for (1), (3), (6)
```
Comment (johnwu93) [inline rfcs/0017-state-snapshotting.md]: Don't think you would want to bin_io a breadcrumb. Seems expensive. We should implement the persistent transition frontier to implement the serialization of breadcrumbs.

File: rfcs/0017-state-snapshotting.md [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+As stated above, snaphotting parts of our application state will give us a
+mechanism for more quickly reproducing edge cases and errors as they arive.
+
+There are several stages. Each of which provides a checkpoint that gives us
+useful features.
+
+[partial]: #partial
+### Partial Snapshotting:
+
+Snapshot enough state to be able to produce the immediate next transition, and
```
Comment (johnwu93) [inline rfcs/0017-state-snapshotting.md]: Producing the immediate next transition means proposing, right? Can you mention that in this paragraph?

File: rfcs/0017-state-snapshotting.md [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+those cases.
+
+[full-offline]: #full-offline
+### Full-offline Snapshotting
+
+Snapshot enough state to recreate a full-node that can continue from the point
+of the snapshot to produce more than just one more transition.
+
+Why is this useful? Sometimes we need more context than the exact block that
+fails. This would give us that context.
```
Comment (johnwu93) [inline rfcs/0017-state-snapshotting.md]: Can you give examples of when we need full-offline snapshotting and full-online snapshotting? For me, I think if I am producing more than one transition, I should be synchronized with the network, right?

File: rfcs/0017-state-snapshotting.md [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+
+I could not find any prior art in the Bitcoin space. It seems like pyethereum
+and parity both support some form of state snapshotting to JSON. It's not clear
+to me that this state captures enough of the information necessary to resume a
+network from that point (and we can't dig through the source due to licensing
+issues), but if it does, it's nice that they managed to use JSON for their
+output format as it's cross-client compatible and easily consumable by other
+tooling.
+
+Some of the data we need to serialize would be a bit hard and slow if we put it
```
Comment (johnwu93) [inline rfcs/0017-state-snapshotting.md]: I've been putting some of our data structures into a json format for our visualizations :D.

File: rfcs/0017-state-snapshotting.md [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+(covered here in detail), and full-online snapshots which is out-of-scope for
+now. Partial snapshotting can be done first and in parallel with implementation
+of transition frontier persistance to make the full-offline version much quicker
+to implement afterwards.
+
+## Motivation
+
+[motivation]: #motivation
+
+As stated above, snaphotting parts of our application state will give us a
```
Comment (cmr) [inline rfcs/0017-state-snapshotting.md]: typo still

**Pull request closed without merging**